kong/clustering/control_plane.lua

```diff
@@ -196,128 +253,271 @@ do
-                  "Sync is suspended for this DP and will resume " ..
-                  "automatically once this DP also upgrades to 2.3 or later"
+  if not major_cp then
+    return nil, "data plane version " .. dp_version .. " is incompatible with control plane version",
+                CLUSTERING_SYNC_STATUS.KONG_VERSION_INCOMPATIBLE
   end
 
-  local major_cp, minor_cp = KONG_VERSION:match(MAJOR_MINOR_PATTERN)
-  local major_node, minor_node = node_version:match(MAJOR_MINOR_PATTERN)
-  minor_cp = tonumber(minor_cp)
-  minor_node = tonumber(minor_node)
-
-  if major_cp ~= major_node or minor_cp - 2 > minor_node or minor_cp < minor_node then
-    return false, "version incompatible, CP version: " .. KONG_VERSION ..
-                  " DP version: " .. node_version ..
-                  " DP versions acceptable are " ..
-                  major_cp .. "." .. math.max(0, minor_cp - 2) .. " to " ..
-                  major_cp .. "." .. minor_cp .. "(edges included)",
-                  CLUSTERING_SYNC_STATUS.KONG_VERSION_INCOMPATIBLE
+  if not major_dp then
+    return nil, "data plane version is incompatible with control plane version " ..
+                KONG_VERSION .. " (" .. major_cp .. ".x.y are accepted)",
+                CLUSTERING_SYNC_STATUS.KONG_VERSION_INCOMPATIBLE
   end
 
-  -- allow DP to have a superset of CP's plugins
-  local p, np
-  local i, j = #self.plugins_list, #node_plugins
+  if major_cp ~= major_dp then
+    return nil, "data plane version " .. dp_version ..
+                " is incompatible with control plane version " ..
+                KONG_VERSION .. " (" .. major_cp .. ".x.y are accepted)",
+                CLUSTERING_SYNC_STATUS.KONG_VERSION_INCOMPATIBLE
+  end
 
-  if j < i then
-    return false, "CP and DP does not have same set of plugins installed",
-                  CLUSTERING_SYNC_STATUS.PLUGIN_SET_INCOMPATIBLE
+  if minor_cp ~= minor_dp then
+    local msg = "data plane minor version " .. dp_version ..
+                " is different to control plane minor version " ..
+                KONG_VERSION
+
+    minor_cp = tonumber(minor_cp, 10)
+    minor_dp = tonumber(minor_dp, 10)
+
+    local difference = math.abs(minor_cp - minor_dp)
+    if difference == 1 then
+      ngx_log(ngx_DEBUG, msg, dp)
+    elseif difference == 2 then
+      ngx_log(ngx_INFO, msg, dp)
+    elseif difference == 3 then
+      ngx_log(ngx_NOTICE, msg, dp)
+    else
+      ngx_log(ngx_WARN, msg, dp)
+    end
   end
 
-  while i > 0 and j > 0 do
-    p = self.plugins_list[i]
-    np = node_plugins[j]
+  for _, plugin in ipairs(self.plugins_list) do
+    local name = plugin.name
+    local cp_plugin = self.plugins_map[name]
+    local dp_plugin = dp_plugin_map[name]
 
-    if p.name ~= np.name then
-      goto continue
+    if not dp_plugin then
+      if cp_plugin.version then
+        ngx_log(ngx_WARN, name, " plugin ", cp_plugin.version, " is missing from data plane", dp)
```

Comment:
generally i would suggest a prefix for those logs for better grokking, like `[clustering]` ?

Reply:
let's emit one type of error in case DP missed it, whether cp_plugin.version is set or not,
how you think? having multiple log
shape for similar outcome will probably makes harder to filter.

Reply:
The prefix is added. There is no difference in log statement especially on single cp or with same version of cps. It is just that other has version and other has not. They always do, so the other is just being defensive.